sync/class.jetpack-sync-module-import.php

```diff
@@ -2,86 +2,153 @@
-
-		$importer_name = $this->get_importer_name( $importer );
-
-		/**
-		 * Sync Event that tells that the import is finished
-		 *
-		 * @since 5.0.0
-		 *
-		 * $param string $importer
-		 */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
-		$this->import_end = true;
-	}
+	public function sync_import_action( $importer ) {
+		$import_action = current_filter();
+		// Map action to event name.
+		$sync_action = self::$import_sync_action_map[ $import_action ];
 
-	public function sync_import_end() {
-		// We already ran an send the import
-		if ( $this->import_end ) {
+		// Only sync each action once per import.
+		if ( array_key_exists( $sync_action, $this->synced_actions ) && $this->synced_actions[ $sync_action ] ) {
 			return;
 		}
 
-		$this->import_end = true;
-		$importer         = 'unknown';
-		$backtrace        = wp_debug_backtrace_summary( null, 0, false );
-		if ( $this->is_importer( $backtrace, 'Blogger_Importer' ) ) {
-			$importer = 'blogger';
-		}
+		// Mark this action as synced.
+		$this->synced_actions[ $sync_action ] = true;
 
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WC_Tax_Rate_Importer' ) ) {
-			$importer = 'woo-tax-rate';
+		// Prefer self-reported $importer value.
+		if ( ! $importer ) {
+			// Fall back to inferring by calling class name.
+			$importer = self::get_calling_importer_class();
 		}
 
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WP_Import' ) ) {
-			// phpcs:ignore WordPress.WP.CapitalPDangit
-			$importer = 'wordpress';
+		// Get $importer from known_importers.
+		if ( isset( self::$known_importers[ $importer ] ) ) {
+			$importer = self::$known_importers[ $importer ];
 		}
 
 		$importer_name = $this->get_importer_name( $importer );
 
-		/** This filter is already documented in sync/class.jetpack-sync-module-posts.php */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
+		do_action( $sync_action, $importer, $importer_name );
 	}
 
 	private function get_importer_name( $importer ) {
 		$importers = get_importers();
 		return isset( $importers[ $importer ] ) ? $importers[ $importer ][0] : 'Unknown Importer';
 	}
 
-	private function is_importer( $backtrace, $class_name ) {
-		foreach ( $backtrace as $trace ) {
-			if ( strpos( $trace, $class_name ) !== false ) {
-				return true;
+	/**
+	 * Determine the class that extends `WP_Importer` which is responsible for
+	 * the current action. Designed to be used within an action handler.
+	 *
+	 * @return string  The name of the calling class, or 'unknown'.
+	 */
+	private static function get_calling_importer_class() {
+		// If WP_Importer doesn't exist, neither will any importer that extends it.
+		if ( ! class_exists( 'WP_Importer' ) ) {
+			return 'unknown';
+		}
+
+		$action    = current_filter();
+		$backtrace = wp_debug_backtrace_summary( null, 0, false );
+
+		$do_action_pos = -1;
+		$backtrace_len = count( $backtrace );
+		for ( $i = 0; $i < $backtrace_len; $i++ ) {
```

Comment:
have you seen https://github.com/Automattic/jetpack/blob/dd9c8a374f4a136e4499853becb124b4270561d3/class.jetpack.php#L6965 maybe we could extend that one or add a new method to Jetpack.class

Reply:
Thanks for the suggestion @lezama. I wasn't aware of that method prior to creating `get_calling_importer_class`. I explored creating something similar in `Jetpack.class`, but there are a few issues with such an approach:

* The existing method simply returns _if_ any of an array of names is a function in the backtrace. It'd be pretty trivial to extend that to support any column of the backtrace (not just `function`, but also `class`), but we care about _which_ class is found in the backtrace.
* We currently care not only about known importers, but any class that extends `WP_Importer`. We could of course add logic similar to the logic we have in `get_calling_importer_class` that checks for subclasses, but that would undo many of the optimizations currently in place in `is_function_in_backtrace`
* All of these extra needs are really very specific to our implementation of getting the calling class in Sync for importers. Until we have the need to do something similar outside of Jetpack Sync, I'm hesitant to write something more generalized without a better idea of how it'd be used.